File: csrc/compiler.cpp
#include "compiler.h"
#include <stdexcept>
#include <map>
#include <iostream>

std::string FactoryCompiler::generate_unique_label()
{
    return std::string("L") + std::to_string(current_label_index++);
}

FactoryCompiler::FactoryCompiler()
    : const_table(new ConstantTable())
{
    _layouts.push(LayoutTracker());
}

void FactoryCompiler::add_op(Opcode op)
{
    code.push_back(OpcodeWithMetadata(std::move(op)));
}

void FactoryCompiler::add_op(OpcodeWithMetadata op)
{
    code.push_back(std::move(op));
}

OpcodeParamType FactoryCompiler::register_const(int64_t value)
{
    return ct_add_int(const_table.get(), value);
}

void FactoryCompiler::compile_expr(const Expression *expr, const std::string &top_label)
{
    const auto need_label = top_label != "";

    switch (expr->type)
    {
    case EXPR_BINARY:
    {
        const Expression *left = expr->expr.bin->left;
        const Expression *right = expr->expr.bin->right;
        OpcodeKind binop;
        switch (expr->expr.bin->op)
        {
        case BINOP_PLUS:
            binop = OPC_ADD2;
            break;
        case BINOP_MINUS:
            binop = OPC_SUB2;
            break;
        case BINOP_TIMES:
            binop = OPC_MUL2;
            break;
        case BINOP_DIVIDE:
            binop = OPC_DIV2;
            break;
        case BINOP_MODULO:
            binop = OPC_MOD2;
            break;
        case BINOP_EQ:
            binop = OPC_EQ2;
            break;
        case BINOP_NEQ:
            binop = OPC_NEQ2;
            break;
        case BINOP_LT:
            binop = OPC_LT2;
            break;
        case BINOP_LE:
            binop = OPC_LE2;
            break;
        case BINOP_GT:
            binop = OPC_GT2;
            break;
        case BINOP_GE:
            binop = OPC_GE2;
            break;
        }
        compile_expr(left, top_label);
        compile_expr(right);
        // now we have the two operands on the stack
        Opcode op;
        op.tag = binop;
        op.param = 0;
        add_op(op);
        break;
    }
    case EXPR_LITERAL:
    {
        switch (expr->expr.lit->type)
        {
        case LIT_INTEGER:
        {
            Opcode op;
            op.tag = OPC_CONST_INT;
            op.param = register_const(expr->expr.lit->value.int_value);
            if (need_label)
            {
                OpcodeWithMetadata::Metadata op_md;
                op_md.this_label = top_label;
                add_op(OpcodeWithMetadata(op, op_md));
            }
            else
            {
                add_op(op);
            }
        };
        }
        break;
    }
    case EXPR_NAME:
    {
        std::string name = expr->expr.name->name;
        uint32_t index = _layouts.top().get_local(name);
        Opcode op;
        op.tag = OPC_LOAD;
        op.param = index;
        if (need_label)
        {
            OpcodeWithMetadata::Metadata op_md;
            op_md.this_label = top_label;
            add_op(OpcodeWithMetadata(op, op_md));
        }
        else
        {
            add_op(op);
        }
        break;
    }
    default:
    {
        throw std::runtime_error("compiler: unimplemented expr type");
        break;
    }
    }
}

void FactoryCompiler::link_jumps()
{
    // first pass: collect all labels and their addresses
    std::map<std::string, uint32_t> labels;
    for (uint32_t i = 0; i < code.size(); i++)
    {
        const auto &op = code[i];
        const auto label = op.get_label();
        if (label != "")
        {
            std::cout << "label: " << label << std::endl;
            labels.insert(std::make_pair(std::string(label), i));
        }
    }

    // second pass: apply the addresses to the jump instructions
    for (uint32_t i = 0; i < code.size(); i++)
    {
        const auto &jmp_to_label = code[i].get_jmp_to_label();
        auto &op = code[i].get_op();
        switch (op.tag)
        {
        case JMP_IF_TRUE:
        case JMP_ALWAYS:
            if (jmp_to_label != "")
            {
                op.param = labels.at(jmp_to_label);
            }
            break;
        default:
            break;
        }
    }
}

void FactoryCompiler::compile_stmt(const Statement *stmt, const std::string &top_label)
{
    const auto need_label = top_label != "";

    switch (stmt->type)
    {
    case STMT_EXPR:
    {
        compile_expr(stmt->stmt.expr);
        Opcode op;
        op.tag = OPC_DISCARD;
        if (need_label)
        {
            OpcodeWithMetadata::Metadata op_md;
            op_md.this_label = top_label;
            add_op(OpcodeWithMetadata(op, op_md));
        }
        else
        {
            add_op(op);
        }
        break;
    }
    case STMT_ASSIGN:
    {
        const char *name = stmt->stmt.assign->name;
        uint32_t assigned_index = _layouts.top().register_local(std::string(name));
        compile_expr(stmt->stmt.assign->expr);
        Opcode op;
        op.tag = OPC_STORE;
        op.param = assigned_index;
        if (need_label)
        {
            OpcodeWithMetadata::Metadata op_md;
            op_md.this_label = top_label;
            add_op(OpcodeWithMetadata(op, op_md));
        }
        else
        {
            add_op(op);
        }
        break;
    }
    case STMT_BLOCK:
    {
        const Statement *orig_stmt = stmt->stmt.blk_start;
        for (const Statement *stmt = orig_stmt; stmt != nullptr; stmt = stmt->blk_next)
        {
            compile_stmt(stmt, top_label);
        }
        break;
    }
    case STMT_CONDITIONAL:
    {
        // first evaluate the condition expression
        compile_expr(stmt->stmt.cond->cond, top_label); // FVAL_BOOL should be on the stack top

        // jump if true
        const auto true_label = generate_unique_label();
        Opcode jmp_true_op;
        jmp_true_op.tag = JMP_IF_TRUE;
        jmp_true_op.param = 0; // will be filled later
        OpcodeWithMetadata::Metadata jmp_true_op_md;
        jmp_true_op_md.jmp_to_label = true_label;
        const auto jmp_true_op_ = OpcodeWithMetadata(jmp_true_op, jmp_true_op_md);
        add_op(jmp_true_op_);

        // jump if false
        std::string false_label = "";
        bool need_false = stmt->stmt.cond->otherwise != NULL;
        if (need_false)
        {
            false_label = generate_unique_label();
            Opcode jmp_false_op;
            jmp_false_op.tag = JMP_ALWAYS;
            jmp_false_op.param = 0; // will be filled later
            OpcodeWithMetadata::Metadata jmp_false_op_md;
            jmp_false_op_md.jmp_to_label = false_label;
            const auto jmp_false_op_ = OpcodeWithMetadata(jmp_false_op, jmp_false_op_md);
            add_op(jmp_false_op_);
        }

        // generate code for the true branch
        compile_stmt(stmt->stmt.cond->then, true_label);

        // generate code for the false branch
        if (need_false)
        {
            compile_stmt(stmt->stmt.cond->otherwise, false_label);
        }
        break;
    }
    default:
    {
        throw std::runtime_error("compiler: unimplemented stmt type");
        break;
    }
    }
}

std::vector<Opcode> FactoryCompiler::get_code()
{
    std::vector<Opcode> result;
    result.reserve(code.size());
    for (const auto &op : code)
    {
        result.push_back(op.get_op());
    }
    return result;
}

const ConstantTable &FactoryCompiler::get_const_table()
{
    return *const_table;
}

LayoutTracker::LayoutTracker()
{
}

uint32_t LayoutTracker::register_local(const std::string &name)
{
    for (const auto &local : locals)
    {
        if (std::get<0>(local) == name)
            return std::get<1>(local);
    }

    uint32_t index = locals.size();
    locals.push_back(std::make_tuple(name, index));
    return index;
}

uint32_t LayoutTracker::get_local(const std::string &name)
{
    for (const auto &local : locals)
    {
        if (std::get<0>(local) == name)
            return std::get<1>(local);
    }

    throw std::runtime_error("local not found");
}

OpcodeWithMetadata::OpcodeWithMetadata(Opcode op, Metadata md)
    : op(op), md(std::move(md))
{
}

OpcodeWithMetadata::OpcodeWithMetadata(Opcode op)
    : op(op), md(Metadata())
{
}

Opcode OpcodeWithMetadata::get_op() const
{
    return op;
}

Opcode &OpcodeWithMetadata::get_op()
{
    return op;
}

const std::string &OpcodeWithMetadata::get_label() const
{
    return md.this_label;
}

const std::string &OpcodeWithMetadata::get_jmp_to_label() const
{
    return md.jmp_to_label;
}

File: src/compiler.rs
use std::collections::HashMap;

use crate::{
    ast::{Expression, LiteralExpression, Statement},
    opcode::Opcode,
};

#[derive(Debug)]
pub struct Compiler {
    code: Vec<OpcodeWithMetadata>,
    layouts: Vec<LayoutTracker>,

    current_label_index: u32,
}

impl Compiler {
    fn add_op(&mut self, op: Opcode) {
        self.code.push(OpcodeWithMetadata::new_op(op));
    }

    fn add_op_md(&mut self, op: Opcode, md: Metadata) {
        self.code.push(OpcodeWithMetadata::new(op, md));
    }

    pub fn new() -> Self {
        Self {
            code: Vec::new(),
            layouts: vec![LayoutTracker::new()],
            current_label_index: 0,
        }
    }

    pub fn compile_expr(&mut self, expr: &Expression, top_label: Option<&str>) {
        match expr {
            Expression::Binary(bin) => {
                let op = bin.op();
                let left = bin.left();
                let right = bin.right();
                // generate operands
                self.compile_expr(left, top_label);
                self.compile_expr(right, None);

                // generate operator
                use crate::ast::BinaryOperator::*;
                let op = match op {
                    Plus => Opcode::Add2,
                    Minus => Opcode::Sub2,
                    Times => Opcode::Mul2,
                    Divide => Opcode::Div2,
                    Modulo => Opcode::Mod2,
                    Eq => Opcode::Eq2,
                    Neq => Opcode::Neq2,
                    Lt => Opcode::Lt2,
                    Le => Opcode::Le2,
                    Gt => Opcode::Gt2,
                    Ge => Opcode::Ge2,
                };
                self.add_op(op);
            }
            Expression::Literal(lit) => match lit {
                LiteralExpression::Integer(i) => {
                    let op = Opcode::ConstInt(*i);
                    match top_label {
                        None => {
                            self.add_op(op);
                        }
                        Some(top_label) => {
                            let md = Metadata {
                                this_label: Some(top_label.to_string()),
                                jmp_to_label: None,
                            };
                            self.add_op_md(op, md);
                        }
                    }
                }
                _ => unimplemented!(),
            },
            Expression::Name(name) => {
                let var_name = name.get_name();
                let var_index = self.layouts.first().unwrap().get_local(var_name).unwrap();
                let op = Opcode::Load(var_index);

                match top_label {
                    None => {
                        self.add_op(op);
                    }
                    Some(top_label) => {
                        let md = Metadata {
                            this_label: Some(top_label.to_string()),
                            jmp_to_label: None,
                        };
                        self.add_op_md(op, md);
                    }
                }
            }
            Expression::FunCall(func) => {
                let callee = func.callee();
                let args = func.args();

                // generate callee
                self.compile_expr(callee, None);

                // generate args (from left to right)
                for arg in args.iter() {
                    self.compile_expr(arg, None);
                }

                // generate call
                let op = Opcode::CallNoKw(args.len());
                self.add_op(op);

                // generate return destination
                self.add_op(Opcode::Nop);
            }
        }
    }

    pub fn compile_stmt(&mut self, stmt: &Statement, top_label: Option<&str>) {
        match stmt {
            Statement::Expression(expr) => {
                self.compile_expr(expr, top_label);
                let discard = Opcode::Discard;
                self.add_op(discard);
            }
            Statement::Assignment(assign) => {
                let name = assign.name();
                let assigned_index = self
                    .layouts
                    .first_mut()
                    .unwrap()
                    .register_local(name.to_string());
                self.compile_expr(assign.expression(), top_label);
                let op = Opcode::Store(assigned_index);
                self.add_op(op);
            }
            Statement::Block(blk) => {
                for (i, stmt) in blk.iter().enumerate() {
                    self.compile_stmt(stmt, if i == 0 { top_label } else { None });
                }
            }
            Statement::Conditional(cond) => {
                // evaluate condition
                let cond_expr = cond.cond();
                self.compile_expr(cond_expr, top_label);

                // jump if true
                let true_label = self.generate_unique_label();
                let true_jmp_op = {
                    let op = Opcode::JmpIfTrue(0);
                    let md = Metadata {
                        this_label: None,
                        jmp_to_label: Some(true_label.clone()),
                    };
                    (op, md)
                };
                self.add_op_md(true_jmp_op.0, true_jmp_op.1);

                // jump if false
                let false_label = match cond.otherwise() {
                    None => None,
                    Some(_) => {
                        let false_label = self.generate_unique_label();
                        let false_jmp_op = {
                            let op = Opcode::JmpAlways(0);
                            let md = Metadata {
                                this_label: None,
                                jmp_to_label: Some(false_label.clone()),
                            };
                            (op, md)
                        };
                        self.add_op_md(false_jmp_op.0, false_jmp_op.1);
                        Some(false_label)
                    }
                };

                // branch end label
                let branch_end_label = self.generate_unique_label();

                // code for true branch
                self.compile_stmt(cond.then(), Some(true_label.as_str()));
                self.add_op_md(
                    Opcode::JmpAlways(0),
                    Metadata {
                        this_label: None,
                        jmp_to_label: Some(branch_end_label.clone()),
                    },
                );

                // code for false branch
                if let Some(false_label) = false_label {
                    self.compile_stmt(cond.otherwise().unwrap(), Some(false_label.as_str()));
                }

                // branch end label
                self.add_op_md(
                    Opcode::Nop,
                    Metadata {
                        this_label: Some(branch_end_label),
                        jmp_to_label: None,
                    },
                );
            }
            Statement::While(wh) => {
                let cond = wh.cond();
                let body = wh.body();

                let cond_label = self.generate_unique_label();
                let body_end_label = self.generate_unique_label();

                // evaluate condition
                self.compile_expr(cond, Some(cond_label.as_str()));
                // jump if condition is not met
                self.add_op_md(
                    Opcode::JmpIfFalse(0),
                    Metadata {
                        this_label: None,
                        jmp_to_label: Some(body_end_label.clone()),
                    },
                );

                // generate body
                self.compile_stmt(body, None);
                // jump back to condition
                self.add_op_md(
                    Opcode::JmpAlways(0),
                    Metadata {
                        this_label: None,
                        jmp_to_label: Some(cond_label.clone()),
                    },
                );

                // end of while
                self.add_op_md(
                    Opcode::Nop,
                    Metadata {
                        this_label: Some(body_end_label),
                        jmp_to_label: None,
                    },
                );
            }
            Statement::FuncDef(def) => {
                let func_name = def.name();
                let func_params = def.params();
                let func_body = def.body();

                let func_def_end_label = self.generate_unique_label();
                self.add_op_md(
                    Opcode::JmpAlways(0),
                    Metadata {
                        this_label: None,
                        jmp_to_label: Some(func_def_end_label.clone()),
                    },
                );

                let func_body_label = self.generate_func_label(func_name);

                self.push_layout();
                {
                    let lay = self.layouts.first_mut().unwrap();
                    for param in func_params.iter() {
                        lay.register_local(param.to_string());
                    }
                }
                self.compile_stmt(func_body, Some(&func_body_label.as_str()));
                self.add_op(Opcode::ConstNull);
                self.add_op(Opcode::Return);
                self.pop_layout();

                self.add_op_md(
                    Opcode::CreateFunction(0, func_params.len()),
                    Metadata {
                        this_label: Some(func_def_end_label),
                        jmp_to_label: Some(func_body_label.clone()),
                    },
                );
                let func_index = self
                    .layouts
                    .first_mut()
                    .unwrap()
                    .register_local(func_name.to_string());
                self.add_op(Opcode::Store(func_index));
            }
        }
    }

    fn generate_unique_label(&mut self) -> String {
        let label = format!("L{}", self.current_label_index);
        self.current_label_index += 1;
        label
    }

    fn generate_func_label(&mut self, func_name: &str) -> String {
        let label = format!("F{}", func_name);
        label
    }

    fn link_jumps(&mut self) {
        let mut label_map: HashMap<String, u32> = HashMap::new();
        // first pass: collect labels and their addresses
        for (i, op) in self.code.iter().enumerate() {
            let label = op.get_label();
            if let Some(label) = label {
                label_map.insert(label.clone(), i as u32);
            }
        }

        // second pass: link jumps
        for op in self.code.iter_mut() {
            let jmp_to_label = op.get_jmp_to_label();
            if let Some(jmp_to_label) = jmp_to_label {
                match op.op {
                    Opcode::JmpIfTrue(_) => {
                        let jmp_to_addr = label_map.get(&jmp_to_label).unwrap();
                        op.op = Opcode::JmpIfTrue(*jmp_to_addr as usize);
                    }
                    Opcode::JmpAlways(_) => {
                        let jmp_to_addr = label_map.get(&jmp_to_label).unwrap();
                        op.op = Opcode::JmpAlways(*jmp_to_addr as usize);
                    }
                    Opcode::JmpIfFalse(_) => {
                        let jmp_to_addr = label_map.get(&jmp_to_label).unwrap();
                        op.op = Opcode::JmpIfFalse(*jmp_to_addr as usize);
                    }
                    Opcode::CreateFunction(_, n) => {
                        let jmp_to_addr = label_map.get(&jmp_to_label).unwrap();
                        op.op = Opcode::CreateFunction(*jmp_to_addr as usize, n);
                    }
                    _ => {}
                }
            }
        }
    }

    pub fn link(&mut self) {
        self.link_jumps();
    }

    pub fn code(&self) -> Vec<Opcode> {
        self.code.iter().map(|op| op.op.clone()).collect()
    }

    fn push_layout(&mut self) {
        self.layouts.push(LayoutTracker::new());
    }

    fn pop_layout(&mut self) {
        self.layouts.pop();
    }
}

#[derive(Debug, Clone)]
struct LayoutTracker {
    locals: Vec<(String, usize)>,
}

impl LayoutTracker {
    pub fn register_local(&mut self, name: String) -> usize {
        // if already exists, return index
        if let Some(i) = self.get_local(&name) {
            return i;
        }

        // otherwise, register and return index
        self.locals.push((name, self.locals.len()));
        self.locals.len() - 1
    }

    pub fn get_local(&self, name: &str) -> Option<usize> {
        for (n, i) in &self.locals {
            if n == name {
                return Some(*i);
            }
        }
        None
    }

    pub fn new() -> Self {
        Self { locals: Vec::new() }
    }
}

#[derive(Debug, Clone)]
pub struct OpcodeWithMetadata {
    op: Opcode,
    md: Metadata,
}

impl OpcodeWithMetadata {
    pub fn new(op: Opcode, md: Metadata) -> Self {
        Self { op, md }
    }

    pub fn new_op(op: Opcode) -> Self {
        Self {
            op,
            md: Metadata {
                jmp_to_label: None,
                this_label: None,
            },
        }
    }

    pub fn get_label(&self) -> Option<String> {
        self.md.this_label.clone()
    }

    pub fn get_jmp_to_label(&self) -> Option<String> {
        self.md.jmp_to_label.clone()
    }
}

#[derive(Debug, Clone)]
pub struct Metadata {
    pub jmp_to_label: Option<String>,
    pub this_label: Option<String>,
}
